### PR TITLE
[interview-practice] mdelcarlo

### PR DESCRIPTION
**Pull request by mdelcarlo**
[interview-practice] mdelcarlo

### CODE REVIEW

File: cracking-the-coding-interview/data-structures/LinkedList.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+      const next = current.next
+      if (next.data === data) {
+        current.next = current.next.next
+        return next
+      }
+    }
+    return
+  }
+
+  length() {
```
Comment (conanbatt) [inline cracking-the-coding-interview/data-structures/LinkedList.ts:62]: This length is O(N), very inefficient!

File: cracking-the-coding-interview/problems/1.strings/1.isUnique.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      return false
+    } else {
+      chars[charAtI] = i
+    }
+  }
+  return true
+}
+
+// What if you cannot use additional data structures ?
+const isUniqueString = (value) => {
```
Comment (conanbatt) [inline cracking-the-coding-interview/problems/1.strings/1.isUnique.ts:19]: The chars variable here is being used as an array. This does not conform to the prompt "do not use additional data structures".

File: cracking-the-coding-interview/problems/1.strings/2.checkPermutation.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+// 2. *Check Permutation*: Given two strings, write a method to decide if one is a permutation of the other. 
+
```
Comment (conanbatt) [inline cracking-the-coding-interview/problems/1.strings/2.checkPermutation.ts:2]: Is O(N log N) the best we can do?

File: cracking-the-coding-interview/problems/5.recusive_and_dinamic_programming/1.tripple_step.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+// 1. * Triple Step *: A child is running up a staircase with n steps and can hop either 1 step, 2 steps, or 3 steps at a time.
+// Implement a method to count how many possible ways the child can run up the stairs.
+// 3 => 3  -  2 1  - 1 2 - 1 1 1
+// 4 => 1111 - 112 - 121 - 211 - 22 - 13 - 31 - 
+
+const countSteps = (numero, memo = {}) => {
+  if(numero === 0) return 1;
+  if (memo[numero]) return memo[numero]
+  const hayTres = numero >= 3
```
Comment (conanbatt) [inline cracking-the-coding-interview/problems/5.recusive_and_dinamic_programming/1.tripple_step.ts:9]: Porque en español? Esto no tiene que aparecer como un habito

File: cracking-the-coding-interview/problems/2.linked_lists/1.removeDups.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+// 1. * Remove Dups *: Write code to remove duplicates from an unsorted linked list.FOLLOW UP
+// How would you solve this problem if a temporary buffer is not allowed ?
+
+const removeDups = (list) => {
+  const setItems = new Set(list);
```
Comment (conanbatt) [inline cracking-the-coding-interview/problems/2.linked_lists/1.removeDups.ts:5]: A set is not an unsorted linked list!

File: cracking-the-coding-interview/problems/2.linked_lists/1.removeDups.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+  return [...setItems.values()]
```
Comment (conanbatt) [inline cracking-the-coding-interview/problems/2.linked_lists/1.removeDups.ts:6]: The input is supposed to be a linked list. Also, the extra prompt is about not using a set.

File: cracking-the-coding-interview/problems/2.linked_lists/7.palindrome.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+const f = new LinkedNode('a');
+
+a.next = b;
+b.next = c;
+c.next = d;
+d.next = e;
+e.next = f;
+
+const checkPalindrome = (node) => {
+  const arr = []
```
Comment (conanbatt) [inline cracking-the-coding-interview/problems/2.linked_lists/7.palindrome.ts:27]: Can we solve the problem without using another data structure? 

I think this is a sliding window challenge.

File: cracking-the-coding-interview/problems/3.stacks_and_queues/2.stackMin.ts [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    delete minStack[popped]
+
+    return popped;
+  }
+
+  this.min = () => {
+    if (array.length === 0) {
+      return undefined; // Stack is empty.
+    }
+    const minValues = Object.values(minStack)
```
Comment (conanbatt) [inline cracking-the-coding-interview/problems/3.stacks_and_queues/2.stackMin.ts:31]: This is not O(1)

File: cracking-the-coding-interview/data-structures/LinkedList.ts [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+  delete(data: T): NodeElement<T> | undefined {
+    if(!this.head) return
+    if(this.head.data === data) {
+      const found = this.head
+      this.head = this.head.next
+      return found
+    }
+    const current: NodeElement<T> | undefined = this.head
+    while (current.next) {
```
[the remark this answers is not in the document]
Comment (mdelcarlo) [reply cracking-the-coding-interview/data-structures/LinkedList.ts:52]: if it does not find it , it never ends

File: cracking-the-coding-interview/problems/1.strings/1.isUnique.ts (hunk 2 above)
Comment (mdelcarlo) [reply cracking-the-coding-interview/problems/1.strings/1.isUnique.ts:19]: fixed

File: cracking-the-coding-interview/problems/1.strings/2.checkPermutation.ts (hunk 3 above)
Comment (mdelcarlo) [reply cracking-the-coding-interview/problems/1.strings/2.checkPermutation.ts:2]: O(n)

File: cracking-the-coding-interview/problems/1.strings/3.URLify.ts [hunk 10]
```diff
@@ -0,0 +1,11 @@
+// 3.  URLify: Write a method to replace all spaces in a string with '%20'.You may assume that the string has sufficient space at the end to hold the additional characters, and that you are given the "true" length of the string.
+
+// ```
+// EXAMPLE
+// Input: "Mr 3ohn Smith "
+// Output: "Mr%203ohn%20Smith"
+// ```
+
+const URLify = (value) => {
```
[the remark this answers is not in the document]
Comment (mdc-nab) [reply cracking-the-coding-interview/problems/1.strings/3.URLify.ts:9]: Its O(n)

File: frontend/pages/challenges/tetris_solution.1.js [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+        setPiece(()=>new Piece())
+      }
+    };
+
+    const movePieceDown = () => {
+      movePiece(0, 1)
+    }
+    const interval = setInterval(movePieceDown, 1000)
+
+    const isColision = (p) => {
```
Comment (conanbatt) [inline frontend/pages/challenges/tetris_solution.1.js:86]: Typo: collision.

File: frontend/pages/challenges/tetris_solution.1.js [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+      for (let i = 0; i < p.shape.length; i++) {
+        for (let j = 0; j < p.shape[0].length; j++) {
+          const isOutOfBoard = (j + p.x < 0) || (i + p.y < 0) || (j + p.x >= board.length)
+          if(isOutOfBoard) return 'out';
```
Comment (conanbatt) [inline frontend/pages/challenges/tetris_solution.1.js:90]: returning booleans or a string is a bit weird...

**Pull request closed without merging**